

```{r custom}
library(VariantAnnotation)
library(ggplot2)
library(devtools)
library(dplyr)
library(pheatmap)
library(scales)
library(gridExtra)
library(gtools)
library(RColorBrewer)

# library(ggbio)
number_ticks <- function(n) {function(limits) pretty(limits, n)}
options(bitmapType = 'cairo')

path_results = "$path_results"
```

```{r create-report, echo=FLASE, eval=FALSE}
library(knitr)
library(rmarkdown)

knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      eval=TRUE, fig.width= 9, echo=FALSE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', bootstrap.show.code=FALSE)
render(file.path(path_results, "report", "report-ready.Rmd"))
```


## quality control metrics

```{r load-metrics}
qc = read.table(file.path(path_results, "metrics", "metrics.tsv"), header=T, sep= "\t", check.names = F)
qc_plus= read.table(file.path(path_results, "basic-bam", "flagstat.tsv"), header=T, sep= "\t", check.names = F)
qc$secondary = unlist(qc_plus[qc_plus$measure=="secondary",as.character(qc$sample)])
qc$singletons = unlist(qc_plus[qc_plus$measure=="singletons",as.character(qc$sample)])
```

### Mean quality along read

```{r read-qual}
qual = read.table(file.path(path_results, "fastqc", "Per_base_sequence_quality.tsv"), header=T, sep= "\t", check.names = F)
qual$sample = as.character(qual$sample)
ggplot(qual, aes(x=Base, y=Mean, group=sample)) +
    geom_line() +
    facet_wrap(~sample) +
    ylim(0,41) +
    scale_color_brewer(palette = "Set1") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5))+
    scale_x_discrete(breaks=number_ticks(10)) 
```

### Size distribution

```{r read-size}
qual = read.table(file.path(path_results, "fastqc", "Sequence_Length_Distribution.tsv"), header=T, sep= "\t", check.names = F)
qual$sample = as.character(qual$sample)

ggplot(qual , aes(x=Length, y=Count, group=sample)) +
    geom_line(size=2, alpha=0.5) +
    scale_color_brewer(palette = "Set1") +
    theme_bw() +
    facet_wrap(~sample) +
    labs(y="# of reads")+
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) +
    scale_x_discrete(breaks=number_ticks(5)) 
```

### Nucleotide content

```{r read-content}
qual = read.table(file.path(path_results, "fastqc", "Per_base_sequence_content.tsv"), header=T, sep= "\t", check.names = F)
qual$sample = as.character(qual$sample)

library(gtools)
qual$Base = factor(qual$Base, levels=mixedsort(unique(qual$Base)))

for (nt in c("G", "A", "T", "C")){
    sample = "sample"
    Base = "Base"
    p = ggplot(qual, aes_string(x=Base, y=nt, group=sample)) +
        geom_line(size=2, alpha=0.5) +
        scale_color_brewer(palette = "Set1") +
        theme_bw() +
        ylab("% of nucleotides") +
        theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust=0.5)) +
        ylim(10,50) + 
        ggtitle(paste0("For nucleotide:",nt))+
    scale_x_discrete(breaks=number_ticks(10)) 
    print(p)
}

```

### Total reads

```{r total-reads}
ggplot(qc, aes(x=sample, y=Total_reads)) + 
    geom_bar(stat = 'identity') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

### Mapped reads

```{r mapped-reads}
ggplot(qc, aes(x=sample, y=Mapped_reads)) + 
    geom_bar(stat = 'identity') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

### off-target reads

```{r off-reads, results='asis', echo=FALSE, eval=FALSE}
tab=data.frame()
for (s in list.files("~/orch/scratch/merck_5_vendors/vendors/work/align")){
    for (fn in list.files(paste0("~/orch/scratch/merck_5_vendors/vendors/work/align/", s), full.names=T))
    if (grepl("offtarget-stats.yaml", fn)){
        d = read.table(fn)
        tab = rbind(tab, data.frame(sample=s, off_reads = d$V2[2]/d$V2[1]*100))
    }
}

ggplot(tab, aes(x=Fixed_name, y=off_reads)) + 
    geom_bar(stat = 'identity') +
    ylab("% off-target reads") +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, colour=col_labs))

```

### Basic metrics by samples

```{r table, results='asis'}
metrics = c("sample", "Total_reads" ,"Mapped_reads_pct", "Duplicates_pct", 
            "singletons", "secondary",
            "%GC", "Sequence_length", "Median_insert_size")
print(kable(qc[, metrics], align="c"))

```

### Variant metrics by samples

```{r table-variants, results='asis'}
qc$ratio_het_hom = qc$Variations_heterozygous/qc$Variations_homozygous
metrics = c("sample", "Variations_total", "Variations_in_dbSNP_pct",
            "Variations_heterozygous", "Variations_homozygous", "ratio_het_hom" )
print(kable(qc[, metrics], align="c"))

```

## coverage

### variants coverage

```{r variants-coverage}
fns = list.files(file.path(path_results, "cg"), full.names = TRUE, pattern = "cg-depth-parse.tsv")
tab = data.frame()
for (fn in fns){
    #print(fn)
    dt = read.table(fn, header=T,sep="\t")
    dt = dt %>% filter(!grepl("[::.::]",depth))
    dt[,2] = as.numeric(dt[,2])
    q = quantile(dt[,2],c(0,.10,.25,.50,.75,.90,1))
    labels=factor(rev(names(q)),levels=c("0%","10%","25%","50%","75%","90%","100%"))
    dt = data.frame(variants_pct=labels, depth=q, sample=dt$sample[1])
    tab = rbind(tab, dt)
}


ggplot(tab, aes(x=depth, y=variants_pct, colour=sample, group=sample)) +
    geom_line(size=2, alpha=.5)+
    theme_bw() +
    labs(list(x="# of reads", y="% variants with more than X reads", title="variants coverage"))
```

### variants coverage vs CG content

```{r variants-coverage-gc, fig.width=15, fig.height=15}
rf <- colorRampPalette(rev(brewer.pal(11,'Spectral')))
r <- rf(32)
list_p = list()
fns = list.files(file.path(path_results, "cg"), full.names = TRUE, pattern = "cg-depth-parse.tsv")
for (fn in fns){
    print(fn)
    dt = read.table(fn, header=T,sep="\t")
    
    dt = dt %>% filter(!grepl("[::.::]",depth))
    dt[,2] = as.numeric(dt[,2])
    sample = dt$sample[1]
    p = ggplot(dt, aes(CG, depth)) +
        stat_bin2d() +
        ylab("# of reads") +
        scale_fill_gradientn(guide = FALSE,colours=r) +
        theme_bw() +
        ylim(0,700) +
        ggtitle(sample)
    
    
    list_p[[as.character(sample)]]=p
    
}

do.call(grid.arrange, list_p)
```

### coverage distribution (completeness)

```{r total-coverage-load}
get_quantile_cov = function(path){
    cov_tab = data.frame()
    co = c('q10', 'q25', 'q50')
    for (fn in list.files(path, pattern = "_cov.tsv", full.names = TRUE)){
            d = read.table(fn, header=T, sep="\t")
            for (i in 1:3){
                # q = quantile(d[,i],c(.10,.25,.50,.75,.90))
                q = quantile(d[,i],c(0.01,.10,.25,.50,.75,.90,.99))
                labels=factor(rev(names(q)),levels=c("1%","10%","25%","50%","75%","90%","99%"))

                s = gsub("-ready","",d[1,6])
                t = data.frame(quantiles=q*100, type=labels, sample=s, min_reads=co[i])
                cov_tab = rbind(cov_tab, t)
            }
        
    }
    cov_tab
}

make_quantile_plots = function(cov_tab){
    p1 = ggplot(cov_tab %>% filter(min_reads=='q10'), aes(x=type, y=quantiles,  group=sample)) +
    geom_line(size=2, alpha=.5)+
    theme_bw()+
    labs(list(x="% of target regions with\nmore than X bases covered", y="% of nt covered\ninside the target", title="considered covered when nt has >10 reads"))

    p2 = ggplot(cov_tab %>% filter(min_reads=='q25'), aes(x=type, y=quantiles,  group=sample)) +
    geom_line(size=2, alpha=.5)+
    theme_bw()+
    labs(list(x="% of target regions with\nmore than X bases covered", y="% of nt covered\ninside the target", title="considered covered when nt has >25 reads"))

    p3 = ggplot(cov_tab %>% filter(min_reads=='q50'), aes(x=type, y=quantiles, group=sample)) +
    geom_line(size=2, alpha=.5)+
    theme_bw()+
    labs(list(x="% of target regions with\nmore than X bases covered", y="% of nt covered\ninside the target", title="considered covered when nt has >50 reads"))
    
    grid.arrange(p1, p2, p3, ncol=1)
}

```

```{r agilent-cov-total-fig, fig.height=12, fig.width=12, cache=TRUE}
cov_tab = get_quantile_cov(file.path(path_results, "coverage"))
make_quantile_plots(cov_tab)
```

### coverage uniformity

Sampling X regions from the target bed file.

```{r cov-uniformity-load}
cov_tab = data.frame()

get_bias_cov = function(path){
    for (fn in list.files(path, pattern = "_bias.tsv", full.names = TRUE)){
            d = read.table(fn, header=T, sep="\t")
            
            cv = d[,"std"]/d[,"mean"]
            bias = (d[,"ntdow"] + d[,"ntup"])/d[,"size"] * 100
            s = as.character(gsub("-ready","",d[1,"sample"]))
            t = data.frame(bias=bias, cv=cv, mean=d[,"mean"], sample=s)
            cov_tab = rbind(cov_tab, t)
            
        
    }
    cov_tab
}

make_bias_plot = function(cov_tab){
    p1 = ggplot(cov_tab, aes(x=log2(mean), y=cv)) +
    geom_point(alpha=0.5) +
    scale_color_brewer(palette = "Set1") + 
    labs(list(y="coefficient of variation",x="log2(mean coverage)")) +
    theme_bw() +
        ggtitle("coverage variation for each target region")

    p2 = ggplot(cov_tab, aes( x=sample,y=bias)) +
    geom_jitter(alpha=0.5) +
    scale_color_brewer(palette = "Set1") + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
    labs(list(y="% of nt with mean-2SD > coverage > mean+2SD ")) +
        ggtitle("% of nucleotides with extreme\ncoverage inside target regions")
    grid.arrange(p1, p2, ncol=1) 
}
```

```{r cov-uniformity-load-agilent, fig.height=12, fig.width=12, cache=TRUE}
bias_tab = get_bias_cov(file.path(path_results,"bias"))
make_bias_plot(bias_tab)    
```

